# 01 — Initial Exploration

## Objective

The goal of this stage is to understand the dataset before making any data cleaning, transformation, or modeling decisions.

We will investigate the dataset's structure, variables, data quality, target distribution, and potential issues that may affect the later stages of the machine learning project.

## Questions

- How large is the dataset?
- What variables are available?
- What does each variable represent?
- What are the actual data types and variable types?
- Are there missing values?
- Are there duplicated records, and how significant are they?
- How is the target variable distributed?
- Are the classes balanced or imbalanced?
- What are the distributions and ranges of the numerical variables?
- How many unique values does each feature contain?
- Which features are binary, ordinal, or continuous?
- Are there any data-quality issues that should be investigated before cleaning?

In [1]:
import numpy as np
import pandas as pd

## 1. Load Data

In this section, we load the raw dataset into a Pandas DataFrame and verify that the data has been loaded correctly.

In [2]:
df = pd.read_csv("../data/raw/diabetes_binary_health_indicators_BRFSS2015.csv")

## 2. Initial Data Overview

In this section, we get a first look at the dataset and examine its dimensions, columns, and sample records.

In [3]:
df.head().T

,0,1,2,3,4
Diabetes_binary,0.0,0.0,0.0,0.0,0.0
HighBP,1.0,0.0,1.0,1.0,1.0
HighChol,1.0,0.0,1.0,0.0,1.0
CholCheck,1.0,0.0,1.0,1.0,1.0
BMI,40.0,25.0,28.0,27.0,24.0
Smoker,1.0,1.0,0.0,0.0,0.0
Stroke,0.0,0.0,0.0,0.0,0.0
HeartDiseaseorAttack,0.0,0.0,0.0,0.0,0.0
PhysActivity,0.0,1.0,0.0,1.0,1.0
Fruits,0.0,0.0,1.0,1.0,1.0


In [4]:
df.shape

(253680, 22)

In [5]:
df.columns

Index(['Diabetes_binary', 'HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker',
       'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies',
       'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth',
       'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education',
       'Income'],
      dtype='str')

## 3. Data Structure & Data Types

In this section, we examine the structure of the dataset, including the number of observations, data types, and non-null values for each feature.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 253680 entries, 0 to 253679
Data columns (total 22 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Diabetes_binary       253680 non-null  float64
 1   HighBP                253680 non-null  float64
 2   HighChol              253680 non-null  float64
 3   CholCheck             253680 non-null  float64
 4   BMI                   253680 non-null  float64
 5   Smoker                253680 non-null  float64
 6   Stroke                253680 non-null  float64
 7   HeartDiseaseorAttack  253680 non-null  float64
 8   PhysActivity          253680 non-null  float64
 9   Fruits                253680 non-null  float64
 10  Veggies               253680 non-null  float64
 11  HvyAlcoholConsump     253680 non-null  float64
 12  AnyHealthcare         253680 non-null  float64
 13  NoDocbcCost           253680 non-null  float64
 14  GenHlth               253680 non-null  float64
 15  MentHlth   

## 4. Missing Values

In this section, we investigate whether the dataset contains missing values and identify which features are affected.

In [7]:
df.isna().sum()

Diabetes_binary         0
HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
dtype: int64

## 5. Duplicate Records

In this section, we investigate the presence of duplicated rows and assess how common they are in the dataset.

In [8]:
df.duplicated().sum()

np.int64(24206)

## 6. Target Variable

In this section, we examine the target variable `Diabetes_binary` and its class distribution.

Understanding the target distribution is important for identifying potential class imbalance before model development.

In [9]:
df.Diabetes_binary.value_counts()

Diabetes_binary
0.0    218334
1.0     35346
Name: count, dtype: int64

## 7. Descriptive Statistics

In this section, we examine summary statistics to understand the ranges, central tendencies, and variability of the dataset's numerical representations.

In [10]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Diabetes_binary,253680.0,0.139333,0.346294,0.0,0.0,0.0,0.0,1.0
HighBP,253680.0,0.429001,0.494934,0.0,0.0,0.0,1.0,1.0
HighChol,253680.0,0.424121,0.494210,0.0,0.0,0.0,1.0,1.0
CholCheck,253680.0,0.962670,0.189571,0.0,1.0,1.0,1.0,1.0
BMI,253680.0,28.382364,6.608694,12.0,24.0,27.0,31.0,98.0
Smoker,253680.0,0.443169,0.496761,0.0,0.0,0.0,1.0,1.0
Stroke,253680.0,0.040571,0.197294,0.0,0.0,0.0,0.0,1.0
HeartDiseaseorAttack,253680.0,0.094186,0.292087,0.0,0.0,0.0,0.0,1.0
PhysActivity,253680.0,0.756544,0.429169,0.0,1.0,1.0,1.0,1.0
Fruits,253680.0,0.634256,0.481639,0.0,0.0,1.0,1.0,1.0


## 8. Feature Cardinality

In this section, we examine the number of unique values in each feature.

This helps us distinguish between binary, ordinal categorical, and continuous numerical variables and better understand how each feature should be treated during preprocessing.

In [11]:
df.nunique().sort_values()

Diabetes_binary          2
HighBP                   2
HighChol                 2
CholCheck                2
Smoker                   2
Stroke                   2
HeartDiseaseorAttack     2
PhysActivity             2
AnyHealthcare            2
Fruits                   2
Veggies                  2
HvyAlcoholConsump        2
NoDocbcCost              2
DiffWalk                 2
Sex                      2
GenHlth                  5
Education                6
Income                   8
Age                     13
MentHlth                31
PhysHlth                31
BMI                     84
dtype: int64

In [12]:
df.nunique().sort_values(ascending=False)

BMI                     84
PhysHlth                31
MentHlth                31
Age                     13
Income                   8
Education                6
GenHlth                  5
Smoker                   2
HighChol                 2
HighBP                   2
Diabetes_binary          2
CholCheck                2
Stroke                   2
HeartDiseaseorAttack     2
NoDocbcCost              2
AnyHealthcare            2
HvyAlcoholConsump        2
Veggies                  2
Fruits                   2
PhysActivity             2
DiffWalk                 2
Sex                      2
dtype: int64

## 9. Feature Understanding

### Target Variable

- **Diabetes_binary**: The target variable indicating whether the respondent has diabetes/prediabetes or not.

### Binary Features

- **HighBP**: Indicates whether the respondent has high blood pressure.
- **HighChol**: Indicates whether the respondent has high cholesterol.
- **CholCheck**: Indicates whether the respondent had their cholesterol checked within the past 5 years.
- **Smoker**: Indicates whether the respondent is a smoker.
- **Stroke**: Indicates whether the respondent has ever had a stroke.
- **HeartDiseaseorAttack**: Indicates whether the respondent has a history of heart disease or heart attack.
- **PhysActivity**: Indicates whether the respondent engages in physical activity.
- **Fruits**: Indicates whether the respondent consumes fruit at least once per day.
- **Veggies**: Indicates whether the respondent consumes vegetables at least once per day.
- **HvyAlcoholConsump**: Indicates whether the respondent is a heavy alcohol consumer.
- **AnyHealthcare**: Indicates whether the respondent has healthcare coverage.
- **NoDocbcCost**: Indicates whether the respondent was unable to see a doctor because of cost.
- **DiffWalk**: Indicates whether the respondent has serious difficulty walking or climbing stairs.
- **Sex**: Binary representation of the respondent's sex.

### Numeric Feature

- **BMI**: Body Mass Index. It is a continuous numerical feature, with values ranging from 12 to 98 in this dataset.

### Ordinal Categorical Features

- **GenHlth**: General health status represented by ordered categories.
- **Age**: Age represented by ordered age categories rather than exact age.
- **Education**: Education level represented by ordered categories.
- **Income**: Income level represented by ordered categories.

### Other Health Features

- **MentHlth**: Number of days during the past 30 days when mental health was not good.
- **PhysHlth**: Number of days during the past 30 days when physical health was not good.

### Initial Observation

Although all features are stored as numerical values (`float64`), they do not all represent continuous numerical variables. Most features are binary, while `Age`, `Education`, `Income`, and `GenHlth` are ordinal categorical variables, and `BMI` is a continuous numerical feature.

## 10. Initial Exploration Findings

### 10.1 What We Learned

- The dataset contains **253,680 observations** and **22 columns**.
- The dataset contains **21 predictor features** and **1 target variable** (`Diabetes_binary`).
- There are **no missing values** in the dataset.
- Most features are represented as binary variables with values of `0` and `1`.
- `BMI` is a continuous numerical feature.
- `Age`, `Education`, `Income`, and `GenHlth` are ordinal categorical features.
- `MentHlth` and `PhysHlth` represent the number of days with poor mental or physical health during the previous 30 days.
- All columns are stored as numerical (`float64`) values, although they represent different types of variables.
- The target variable is binary:
  - `0` → No diabetes
  - `1` → Diabetes / prediabetes.
- The target classes are imbalanced, with approximately **86.07% class 0** and **13.93% class 1**.

### 10.2 Data Quality Observations

The initial exploration revealed several points that require further investigation before preprocessing and model development:

- The dataset contains **24,206 duplicated rows**. We need to determine whether these represent true duplicate records or valid observations with identical feature values before deciding whether they should be removed.
- The target variable is **imbalanced**, with the positive class representing only about 13.93% of the observations. This will need to be considered during model evaluation.
- All features are stored as numerical values, but their actual meanings and statistical types differ. Therefore, preprocessing should not treat all columns in the same way.
- Some variables are ordinal categorical features encoded numerically. We need to verify their value meanings and ordering before applying transformations.
- The dataset contains no missing values, so missing-value handling will not be required unless additional data-quality issues are discovered.
- The observed value ranges and category representations should be checked for consistency and validity before modeling.

### 10.3 Next Step

The next stage will focus on **Data Cleaning & Quality Assessment**, where the observations above will be investigated and appropriate preprocessing decisions will be made based on the evidence found in the data.